v0.14.0 Updates (#121)
dash-docs/dash-docs: _includes/devdoc/guide_dash_features.md

| **LLMQ Signing Sessions**   |   |                         | Quorums internally process locking |
|                             |   |                         | Quorum(s) responsible for the transaction's inputs lock the inputs via LLMQ signing sessions
|                             |   |                         | Once all inputs are locked, the quorum responsible for the overall transaction creates the transaction lock (`islock`) via an LLMQ signing session
| **LLMQ Results**             |   |                         | Quorum results broadcast to the network |
|                             | ← | `inv` message (islock)  | Quorum responds with lock inventory
| `getdata` message (islock)  | → |                         | Client requests lock message
|                             | ← | `islock` message        | Quorum responds with lock message

{% endautocrossref %}

### ChainLocks
{% include helpers/subhead-links.md %}

{% autocrossref %}

Dash's ChainLock feature leverages [LLMQ Signing Requests/Sessions](#llmq-signing-session)
to reduce uncertainty when receiving funds and remove the possibility of 51%
mining attacks.

> For each block, an LLMQ of a few hundred masternodes (300-400) is selected and each

Reviewer: I thought it is exactly 400 MNs

Author: That's the ideal case where there aren't any bad/non-participating members in the DKG setup. It can go as low as 300 though.
https://github.com/dashpay/dash/blob/develop/src/chainparams.cpp#L149-L151
```
        .name = "llmq_400_60",
        .size = 400,
        .minSize = 300,
        .threshold = 240,
```

Reviewer: Even if some MNs are bad/non participating, aren't they technically in the LLMQ?

Author: They would only be in the LLMQ if they participated in DKG. If they don't participate in DKG, they have no contribution to the quorum's key (and therefore aren't part of the quorum). 
@codablock, Can you confirm what I said is correct?

Reviewer: Ahh okay, that seems to make some sense. And if the minimum song participate in DKG then the LLMQ is invalid/not used etc

Author: Yes, if there are less than 300 good members (for `llmq_400_60`), the "candidate quorum" will not be able to form a key and thus not really exist.

Reviewer 2: Confirming what @thephez wrote.

participating member signs the first block that it sees extending the active
chain at the current height. If enough members (at least 240) see the same block
as the first block, they will be able to create a `clsig` message and propagate
it to all nodes in the network.

If a valid `clsig` message is received by a node, it must reject all blocks (and
any descendants) at the same height that do not match the block specified in the
`clsig` message. This makes the decision on the active chain quick, easy and
unambiguous. It also makes reorganizations below this block impossible.

When LLMQ-based InstantSend is enabled, a ChainLock is only attempted once all
transactions in the block are locked via InstantSend. If a block contains

> unlocked transactions, retroactive InstantSend locks are established prior to

Reviewer: Are these retroactive locks in 0.14? I remember some talk about it not making it in.

Author: Initially I think there was some question of whether or not they would be in it, but they were added in https://github.com/dashpay/dash/pull/2770.

{% endautocrossref %}

#### LLMQ InstantSend
{% include helpers/subhead-links.md %}

{% autocrossref %}

The introduction of Long-Living Masternode Quorums in Dash Core 0.14 provides
a foundation to further scale InstantSend. LLMQ-based InstantSend removes a
number of previously required limitations and simplifies the process by decreasing

> the number of P2P messages clients must use.

Reviewer: in the old system the client still only used one message, `ix` now just `tx` but each `ix` message resulted in each MN in the quorum sending a message around which resulted in a lot of traffic. 
```suggestion
the number of P2P messages sent around the network.
```

Author: This sentence was referring to `1 ix + 10*input_count txlvotes` vs `1 tx + 1 islock` (i.e. regular nodes only get an islock and don't have to track all the votes). I'll make some adjustments to mention the network aspect too.

During the evaluation and transition from standard InstantSend to LLMQ-based
InstantSend, Sporks 2 (`SPORK_2_INSTANTSEND_ENABLED`) and 20 (`SPORK_20_INSTANTSEND_LLMQ_BASED`)
will both be used. Spork 2 enables or disables the entire InstantSend feature,
while spork 20 determines which of the two InstantSend mechanisms is active when
InstantSend is enabled.

There are still some limitations on LLMQ-based InstantSend transactions:

* Transaction inputs must either:

>   * Be in a block that has a ChainLock

Reviewer: ```suggestion
    * Be locked by InstantSend
    * Be in a block that has a ChainLock
```

Reviewer: I see it's addressed below, but it makes sense to be in this list. Might want to remove or change line 144 and add this?

  * Have at least the number confirmations (block depth) indicated by the table below

  | **Network** | **Confirmations Required** |
  |---------|--------------|
  | Mainnet | 6 Blocks |
  | Testnet | 2 Blocks |
  | Regtest | 2 Blocks |
  | Devnet  | 2 Blocks |

Improvements from the old InstantSend system:

* Changed: Transactions can be chained if the inputs are from transactions that are also locked
* Changed: InstantSend locks are attempted for all transactions (`tx` messages) - no need to request it via the special message (`ix` message)
* Changed: Only need to receive a single `islock` message - no need to track votes (`txlvote` messages) for each input

> * Removed: Limit on number of inputs

Reviewer: Don't love the change of prefix, I would prefer consistent prefix but because I ended up not reading it by the fourth line so didn't see it changed and was confused :confused:

Author: Yes, it is a little awkward. What do you think about removing the prefix and just doing this:

* Transactions can now be chained if the inputs are from transactions that are also locked
* InstantSend locks are attempted for all transactions (`tx` messages) - no need to request it via the special message (`ix` message)
* Only need to receive a single `islock` message - no need to track votes (`txlvote` messages) for each input
* Limit on number of inputs removed
* Limit on transaction value removed
* Timeout for locking removed - transaction locks will only be removed once the transaction is confirmed in a ChainLocked block
* Custom InstantSend fee remove

Reviewer: I like it :)

Author: I actually changed it to be a table - https://github.com/dash-docs/dash-docs/blob/204d9def1af48b9a7d1b25ce612709aa41bb391c/_includes/devdoc/guide_dash_features.md#llmq-instantsend

already voted on a superblock trigger).

At the superblock height, the trigger with the most "Yes" votes is paid out by
that block's miner.

##### Sentinel<!--noref--> Ping
{% include helpers/subhead-links.md %}

![Warning icon](/img/icons/icon_warning.svg) NOTE: Deprecated by [Dash Core v0.14](https://github.com/dashpay/sentinel<!--noref-->/pull/64)

In Dash Core 12.2, use of the `watchdog` governance object type was replaced
by integrating sentinel information into the masternode ping (`mnp` message)
via [Pull Request #1491](https://github.com/dashpay/dash/pull/1491).

From Dash Core 0.12.2 through Dash Core 0.13, Sentinel used the `sentinelping` RPC
to update the masternode info and prevent it from entering a `MASTERNODE_WATCHDOG_EXPIRED` state.

{% endautocrossref %}

### Masternode Quorums
{% include helpers/subhead-links.md %}

Dash's masternode quorums are used to facilitate the operation of masternode provided
features in a decentralized, deterministic way. The original quorums (used
largely for InstantSend and masternode payments) were ephemeral and used for a
single purpose (e.g. voting on one specific InstantSend transaction).

> Dash Core 0.14 (protocol version 70214) introduced the Long Living Masternode

Reviewer: sounds less weird to me...
```suggestion
Dash Core 0.14 (protocol version 70214) introduced Long Living Masternode
```
